src/clean.c

```diff
@@ -0,0 +1,397 @@
+		return false;
+	}
+	return !memcmp(entry->d_name, prefix, prefix_len);
+}
+
+static char *read_mom_contents(int version)
+{
+	char *mom_path = NULL;
+	char *contents = NULL;
+
+	string_or_die(&mom_path, "%s/%d/Manifest.MoM", state_dir, version);
+	FILE *f = fopen(mom_path, "r");
+	if (!f) {
+		/* This is a best effort. */
+		goto end;
+	}
+
+	int ret;
+	ret = fseek(f, 0, SEEK_END);
+	if (ret != 0) {
+		goto close_then_end;
+	}
+
+	long size = ftell(f);
+	/* Don't care about failures or empty file. */
+	if (size < 1) {
+		goto end;
+	}
+
+	ret = fseek(f, 0, SEEK_SET);
+	if (ret != 0) {
+		goto close_then_end;
+	}
+
+	char *tmp = malloc(size + 1);
+	if (!tmp) {
+		goto close_then_end;
+	}
+
+	ret = fread(tmp, size, 1, f);
+	if (ret != 1) {
+		free_string(&tmp);
+		goto close_then_end;
+	}
+
+	contents = tmp;
+
+close_then_end:
+	fclose(f);
+
+end:
+	free_string(&mom_path);
+	return contents;
+}
+
+static int clean_staged_manifests()
+{
+	int ret = 0;
+	DIR *dir;
+
+	dir = opendir(state_dir);
+	if (!dir) {
+		ret = errno;
+		goto end;
+	}
+
+	/* NOTE: Currently Manifest files have their timestamp from generation
+	 * preserved. */
+
+	/* When --all is not used, keep the Manifests used by the current OS version. This
+	 * ensures that a regular 'clean' won't make 'search' redownload files. */
+	char *mom_contents = NULL;
+	if (!options.all) {
+		int current_version = get_current_version(path_prefix);
+		if (current_version < 0) {
+			fprintf(stderr, "Unable to determine current OS version\n");
+		} else {
+			mom_contents = read_mom_contents(current_version);
+		}
+	}
+
+	while (true) {
+		/* Reset errno to properly identify the end of stream. */
+		errno = 0;
+		struct dirent *entry;
+		entry = readdir(dir);
+		if (!entry) {
+			if (errno) {
+				ret = errno;
+			}
+			break;
+		}
+
+		const char *name = entry->d_name;
+		if (!strcmp(name, ".") || !strcmp(name, "..")) {
+			continue;
+		}
+		if (!is_all_digits(name)) {
+			continue;
+		}
+
+		/* This is not precise: it may keep Manifest files that we don't use, and
+		 * also will keep the previous version. If that extra precision is
+		 * required we should parse the manifest. */
+		if (mom_contents && strstr(mom_contents, name)) {
+			continue;
+		}
+
+		char *version_dir;
+		string_or_die(&version_dir, "%s/%s", state_dir, name);
+		ret = remove_if(version_dir, is_manifest);
+
+		/* Remove empty dirs if possible. */
+		(void)rmdir(version_dir);
+
+		free_string(&version_dir);
+		if (ret != 0) {
+			goto end;
+		}
+	}
+
+end:
+	free_string(&mom_contents);
+	closedir(dir);
+	return ret;
+}
+
+int clean_main(int argc, char **argv)
+{
+	copyright_header("clean");
+
+	if (!parse_options(argc, argv)) {
+		return EXIT_FAILURE;
+	}
+
+	int ret = 0;
+	int lock_fd = 0;
+	ret = swupd_init(&lock_fd);
+	if (ret != 0) {
+		fprintf(stderr, "Failed swupd initialization, exiting now.\n");
+		return ret;
+	}
+
+	ret = clock_gettime(CLOCK_REALTIME, &now);
+	if (ret != 0) {
+		perror("error reading time");
```

Comment:
A better error message would help here. If we have --all do we care if we can't read the time?

Reply:
Changed to only calculate 'now' if we need. Changed the error message.